### DISCUSSION TITLE
I want to make a programming language that can harness GPU and is easier than C can embed in C code can export C code and is modular suitable for big projects.. Any idea?

### DISCUSSION THREAD

**Discussion by mostafa-technet** [community/community#140283] [Programming Help]
### Select Topic Area

Question

### Body

It will be event driven it will have easy multi threading with run async and automatic parallelization (you will tell how many threads).. It makes it more modular for big project like you don't do memory allocation or freeing in other functions in a module etc.. It will also have collections based on trees that work by bitwise operations possible to do in CPU or GPU (I have created algorithms).. 

Some codes I've written in C:

#define NEWGETARRAYVAR(NAME, THREADID, INDEX) (NAME[THREADID][INDEX])
#define NEWSETARRAYVAR(NAME, THREADID, INDEX, VARIABLE) (NAME[THREADID][INDEX] = VARIABLE)
#define NEWGETHEAPVAR(NAME, THREADID, TYPE) (TYPE)NEWGETARRAYVAR(&NAME, THREADID, 0)
#define NEWGETSTACKVAR(NAME, THREADID, TYPE) (TYPE)(NAME[THREADID])
#define NEWSETSTACKVAR(NAME, THREADID, VARIABLE)  (NAME[THREADID] = VARIABLE)
#define NEWSETHEAPVAR(NAME, THREADID, VARIABLE) (NAME[THREADID] = VARIABLE)
#define _UNDEFINED -1256
void* rt[THREADCOUNT];
struct functionStr {
    void** (*f)(void*, ...);    
    void* arg0;
    va_list arg1;    
};

struct functionStr FunCaller(int a, char* RType, ...) {
   
    struct  functionStr fs;
    
    va_list vl, vl2;    
    va_start(vl, a);
    
   
    void** (*f)(void*,...);
    f = va_arg(vl, void*); 

    va_copy(vl2, vl);
    /*for (int i = 1; i <= a; i++) {
        vii[i - 1] = va_arg(vl, void*);
    }*/
    va_end(vl); 
    fs.f = f;
    fs.arg0 = a;
    fs.arg1 = vl2;
    return fs;
}

main()
  apr_initialize();

  apr_pool_t* pool;
  atexit(apr_terminate);
  apr_pool_create(&pool, NULL);
  apr_pool_clear(pool);
  NEWVARIABLE_ON_STACK(int, a);
  NEWSERSTACKVAR(a, threadId, 10);
  NEWVARIABLE_ON_HEAP(int, b);
  if (NEWGETHEAPVAR(b, threadId, void*) != NULL) {
  NEWSETHEAPVAR(b,threadId, 13);
  time_t tm1 = time(NULL);int j4 = 0;
  uv_thread_t tt[THREADCOUNT];

**Comment by LiteBrite82** [comment]
Thanks for posting in the GitHub Community, @mostafa-technet ! 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.

**Reply by mostafa-technet** [reply]
thanks..
